### DISCUSSION TITLE
NodeSet updater?

### DISCUSSION THREAD

**Discussion by swamper123** [FreeOpcUa/opcua-asyncio#357] [Q&A]
Imagine, we are running a server. Used Nodeset in that case would be, let's say uri ".../UA", version = "1.0.0" and a PublicatedDate.

Now we want to bump the same uri Nodeset to version 1.1, via xmlimport.

Do you know, what the wanted behaviour would be? Is that allowed with a running server?
~~I guess new Nodes will be added normally, but what with already existing Nodes?
Should they be overwritten (which cause dataloss?) or the value stored anywhere else and set again when the new Node is created? Or do we ignore them?~~

Reason is, that I am thinking about a better concept for the xmlimporter method (see #353 ), checking for RequiredModels and how to handle it with already existing Namespaces in the server...and perhaps thinking for versions where you pass a list of paths to xmlpaths which would need even more brainpower.  (Which would end up to: checking all xml files of requirements, try to figure out latest versions of all requirements, figure out a good order to import them without conflicts(?) , checking what already exists on the server, compare if the requirement is one of the xml files, recognising missing requirements, watch out for raceconditions...😅 ).

**Comment by AndreasHeine** [comment]
in the machinery nodeset is a required field which tells you what is needed

```xml
  <Models>
    <Model ModelUri="http://opcfoundation.org/UA/Machinery/" Version="1.0.0" PublicationDate="2020-09-25T00:00:00Z">
      <RequiredModel ModelUri="http://opcfoundation.org/UA/" Version="1.04.7" PublicationDate="2020-07-15T00:00:00Z" />
      <RequiredModel ModelUri="http://opcfoundation.org/UA/DI/" Version="1.02.2" PublicationDate="2020-06-02T00:00:00Z" />
    </Model>
  </Models>
```

so probably we neet to check that before start parsing!?

**Reply by swamper123** [reply]
Yes I know that field and able to get access to. I know that there are Nodes in Server->Namespaces to check ModelUri, Version and PublicationDate.

But what if version 1.0 of DI is existing on the Server and you want to bump it to version 1.1?
Should another Namespace being created or the old one replaced/updated?
And how we may protect the process, while Nodes are getting updated?

**Comment by jamilraichouni** [comment]
Sounds a bit related to [an issue](https://github.com/FreeOpcUa/opcua-asyncio/issues/1922) I just opened.
